### ISSUE TITLE
Can't set focus before Application.Run - add TopLevel.Running event

### ISSUE THREAD

**Issue by tig** [migueldeicaza/gui.cs#445]
One would think that you could call `SetFocus` when creating controls, before calling `Application.Run`, but it doesn't work:

```csharp
			_top.SetFocus (_rightPane);
			Application.Run (_top);

```

I found a hacky workaround by using `Application.Iteration`:

```csharp
...
			// HACK: There is no other way to SetFocus before Application.Run
			Application.Iteration += Application_Iteration;
			Application.Run (_top);
		}

		private static void Application_Iteration (object sender, EventArgs e)
		{
			Application.Iteration -= Application_Iteration;
			_top.SetFocus (_rightPane);
		}
```

I propose adding an `Running` event to `TopLevel`:

```csharp
...
			_top.Running += (o, e) => _top.SetFocus (_rightPane);
			Application.Run (_top);
		}
```

I think it would be best called once the `MainLoop` starts (not before). Like this:

```csharp
		public static void RunLoop (RunState state, bool wait = true)
		{
			if (state == null)
				throw new ArgumentNullException (nameof (state));
			if (state.Toplevel == null)
				throw new ObjectDisposedException ("state");

			bool firstIteration = true;
			for (state.Toplevel.Running = true; state.Toplevel.Running;) {
				if (MainLoop.EventsPending (wait)) {

					// Notify Toplevel it's running
					if (firstIteration)
						state.Toplevel?.Invoke(null, EventArgs.Empty);
					firstIteration = false;

					MainLoop.MainIteration ();
					Iteration?.Invoke (null, EventArgs.Empty);
				} else if (wait == false)
					return;
				if (!state.Toplevel.NeedDisplay.IsEmpty || state.Toplevel.childNeedsDisplay) {
					state.Toplevel.Redraw (state.Toplevel.Bounds);
					if (DebugDrawBounds)
						DrawBounds (state.Toplevel);
					state.Toplevel.PositionCursor ();
					Driver.Refresh ();
				} else
					Driver.UpdateCursor ();
			}
		}
```

A simpler implementation would just put it in `Begin`:

```csharp
		static public RunState Begin (Toplevel toplevel)
		{
			if (toplevel == null)
				throw new ArgumentNullException (nameof (toplevel));
			var rs = new RunState (toplevel);

			Init ();
			if (toplevel is ISupportInitializeNotification initializableNotification &&
			    !initializableNotification.IsInitialized) {
				initializableNotification.BeginInit();
				initializableNotification.EndInit();
			} else if (toplevel is ISupportInitialize initializable) {
				initializable.BeginInit();
				initializable.EndInit();
			}
			toplevels.Push (toplevel);
			Current = toplevel;
			Driver.PrepareToRun (MainLoop, ProcessKeyEvent, ProcessKeyDownEvent, ProcessKeyUpEvent, ProcessMouseEvent);
			if (toplevel.LayoutStyle == LayoutStyle.Computed)
				toplevel.RelativeLayout (new Rect (0, 0, Driver.Cols, Driver.Rows));
			toplevel.LayoutSubviews ();
			OnLoad?.Invoke ();
			toplevel.WillPresent ();
			Redraw (toplevel);
			toplevel.PositionCursor ();
			Driver.Refresh ();

			// Notify Toplevel it's running
			toplevel?.Invoke(null, EventArgs.Empty);

			return rs;
		}
```

**Comment by BDisp** [comment]
You can change the OnLoad for that without having multiple actions to invoke. Remember that are static actions, so the last caller supersede the others.

**Comment by BDisp** [comment]
I think the best solution would be to implement `ISupportInitializeNotification` and `ISupportInitialize` in `Toplevel` and its derivatives.

```
if (toplevel is ISupportInitializeNotification initializableNotification &&
	!initializableNotification.IsInitialized) {
	initializableNotification.BeginInit();
	initializableNotification.EndInit();
} else if (toplevel is ISupportInitialize initializable) {
	initializable.BeginInit();
	initializable.EndInit();
}

```

**Comment by tig** [comment]
Why would implementing an interface be better? This is far more cumbersome.

**Comment by BDisp** [comment]
It's in the Core for some reason. So there could be personalized notification for each top-level regardless of where the call was made, as it would be more object oriented. I agree that it would be more complicated but more rationally functional.

**Comment by tig** [comment]
I still don't see the value in adding all the complexity of an interface. This is a notification that simple apps want to receive. Perhaps, under the covers there could be an interface. 

Another reason for not doing it with an interface: There are other notifications we're going to want over time. If we do each one via implementing an interface, before long `Toplevel` will end up looking like this:

     public Toplevel : View, ISupportInitializeNotification, IInterface2, IInterface3, ...

My current proposal is below. 

```csharp
public class Toplevel : View {
		/// <summary>
		/// Gets or sets whether the <see cref="T:Terminal.Gui.MainLoop"/> for this <see cref="T:Terminal.Gui.Toplevel"/> is running or not. Setting
		/// this property to false will cause the MainLoop to exit. 
		/// </summary>
		public bool Running { get; set; }

		/// <summary>
		/// Fired once the Toplevel's <see cref="T:Terminal.Gui.MainLoop"/> has started it's first iteration. 
		/// Subscribe to this event to perform tasks when the <see cref="T:Terminal.Gui.Toplevel"/> has been laid out and focus has been set.
		/// changes. A Ready event handler is a good place to finalize initialization after calling `<see cref="T:Terminal.Gui.Application.Run"/>(topLevel)`. 
		/// </summary>
		public event EventHandler Ready;

		/// <summary>
		/// Called from Application.RunLoop after the Toplevel has entered it's first iteration of the loop. 
		/// </summary>
		internal virtual void OnReady ()
		{
			Ready?.Invoke (this, EventArgs.Empty);
		}

...

		public static void RunLoop (RunState state, bool wait = true)
		{
			if (state == null)
				throw new ArgumentNullException (nameof (state));
			if (state.Toplevel == null)
				throw new ObjectDisposedException ("state");

			bool firstIteration = true;
			for (state.Toplevel.Running = true; state.Toplevel.Running;) {
				if (MainLoop.EventsPending (wait)) {
					// Notify Toplevel it's ready
					if (firstIteration) {
						state.Toplevel.OnReady ();
					}
					firstIteration = false;

					MainLoop.MainIteration ();
					Iteration?.Invoke (null, EventArgs.Empty);
				} else if (wait == false)
					return;
				if (!state.Toplevel.NeedDisplay.IsEmpty || state.Toplevel.childNeedsDisplay) {
					state.Toplevel.Redraw (state.Toplevel.Bounds);
					if (DebugDrawBounds)
						DrawBounds (state.Toplevel);
					state.Toplevel.PositionCursor ();
					Driver.Refresh ();
				} else
					Driver.UpdateCursor ();
			}
		}
```

Example usage:

```csharp
	_top.Ready += (o, e) => _top.SetFocus (_rightPane);
```

I've implemented this and tested it for my scenario and it works dandy.

**Comment by BDisp** [comment]
It's looks good since it only will be raised in the first interaction of that top-level only. Great.

**Comment by tig** [comment]
Closed with #446
